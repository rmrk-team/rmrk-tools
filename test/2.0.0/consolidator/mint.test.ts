import { Consolidator, NFT } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock,
  mintNftMock2,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";

beforeAll(async () => {
  return await cryptoWaitReady();
});

describe("rmrk2.0.0 Consolidator: MINT", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
    ...getBlockCallsMock(mintNftMock().mint()),
    ...getBlockCallsMock(mintNftMock2().mint()),
  ];

  it("Should mint a NFT and make another account owner", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(createCollectionMock().create()),
      ...getBlockCallsMock(mintNftMock().mint(getBobKey().address)),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should mint a NFT and make another NFT an owner", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(createCollectionMock().create()),
      ...getBlockCallsMock(mintNftMock2().mint()),
      ...getBlockCallsMock(mintNftMock().mint(mintNftMock2(4).getId())),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should not allow to mint a NFT without a calss", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(mintNftMock().mint()),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should allow to mint NFT with royalties", async () => {
    const nft = new NFT({
      block: 0,
      collection: createCollectionMock().id,
      symbol: "KANR",
      sn: "999".padStart(8, "0"),
      transferable: 1,
      owner: getBobKey().address,
      properties: {
        royaltyInfo: {
          type: "royalty",
          value: {
            royaltyPercentFloat: 2.3,
            receiver: getBobKey().address,
          },
        },
      },
    });

    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(createCollectionMock().create()),
      ...getBlockCallsMock(nft.mint()),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should not allow to mint NFT with invalid royalties", async () => {
    const nft = new NFT({
      block: 0,
      collection: createCollectionMock().id,
      symbol: "KANR",
      sn: "999".padStart(8, "0"),
      transferable: 1,
      owner: getBobKey().address,
      properties: {
        royaltyInfo: {
          type: "royalty",
          value: {
            royaltyPercentFloat: 101,
            receiver: getBobKey().address,
          },
        },
      },
    });

    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(createCollectionMock().create()),
      ...getBlockCallsMock(nft.mint()),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });
});
