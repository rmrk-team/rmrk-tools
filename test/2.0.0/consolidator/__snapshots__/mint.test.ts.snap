// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: MINT Should allow to mint NFT with royalties 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000999": Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "royaltyInfo": Object {
          "type": "royalty",
          "value": Object {
            "receiver": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
            "royaltyPercentFloat": 2.3,
          },
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000999",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: MINT Should mint a NFT and make another NFT an owner 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888": Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777": Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "pending": true,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: MINT Should mint a NFT and make another account owner 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777": Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: MINT Should not allow to mint NFT with invalid royalties 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  },
  "invalid": Array [
    Object {
      "block": 3,
      "caller": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "message": "[MINT] Dead before instantiation: Royalty percentage value have to be between 0 and 100",
      "object_id": "RMRK::MINT::2.0.0::{\\"collection\\"%3A\\"d43593c715a56da27d-KANARIABIRDS\\"%2C\\"symbol\\"%3A\\"KANR\\"%2C\\"transferable\\"%3A1%2C\\"sn\\"%3A\\"00000999\\"%2C\\"properties\\"%3A{\\"royaltyInfo\\"%3A{\\"type\\"%3A\\"royalty\\"%2C\\"value\\"%3A{\\"royaltyPercentFloat\\"%3A101%2C\\"receiver\\"%3A\\"5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty\\"}}}}",
      "op_type": "MINT",
    },
  ],
  "nfts": Object {},
}
`;

exports[`rmrk2.0.0 Consolidator: MINT Should not allow to mint a NFT without a calss 1`] = `
Object {
  "bases": Object {},
  "collections": Object {},
  "invalid": Array [
    Object {
      "block": 2,
      "caller": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "message": "NFT referencing non-existant parent collection d43593c715a56da27d-KANARIABIRDS",
      "object_id": "2-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "op_type": "MINT",
    },
  ],
  "nfts": Object {},
}
`;
